app/controllers/oauth_controller.rb

```diff
@@ -0,0 +1,55 @@
+class OauthController < ApplicationController
+  FRONTEND_URL = ENV['FRONTEND_URL'].freeze
+  skip_after_action :cors_set_access_control_headers, only: :access_token
+
+  def request_token
+    settings = params[:oauth_settings]
+    consumer = create_consumer(settings)
+
+    request_token = consumer.get_request_token(oauth_callback: "#{oauth_access_token_url}/?instance_id=#{params[:instance_id]}")
+
+    session[:oauth] = settings.to_hash
+    session[:oauth][:oauth_token] = request_token.token
+    session[:oauth][:oauth_token_secret] = request_token.secret
+
+    render json: { authorize_url: request_token.authorize_url }
+  end
+
+  def access_token
+    settings = session[:oauth].to_hash.with_indifferent_access
+    session[:oauth] = nil
+
+    consumer = create_consumer(settings)
+    request_token = OAuth::RequestToken.new(consumer, settings[:oauth_token], settings[:oauth_token_secret])
+    access_token = request_token.get_access_token(oauth_verifier: params[:oauth_verifier])
+
+    # Can't return consumer secret because it's too long for the query string
```

Comment:
I would return it so they don't have to type it in the settings again. You can't save your settings without an access token and secret first in Rainforest.